triedb/pathdb/history_trienode.go

```diff
@@ -0,0 +1,732 @@
+	nodeList := make(map[common.Hash][]string)
+	for owner, subset := range nodes {
+		keys := sort.StringSlice(slices.Collect(maps.Keys(subset)))
+		keys.Sort()
+		nodeList[owner] = keys
+	}
+	return &trienodeHistory{
+		meta: &trienodeMetadata{
+			version: trienodeHistoryVersion,
+			parent:  parent,
+			root:    root,
+			block:   block,
+		},
+		owners:   slices.SortedFunc(maps.Keys(nodes), common.Hash.Cmp),
+		nodeList: nodeList,
+		nodes:    nodes,
+	}
+}
+
+// sharedLen returns the length of the common prefix shared by a and b.
+func sharedLen(a, b []byte) int {
+	n := len(a)
+	if len(b) < n {
+		n = len(b)
+	}
+	for i := 0; i < n; i++ {
+		if a[i] != b[i] {
+			return i
+		}
+	}
+	return n
+}
+
+// typ implements the history interface, returning the historical data type held.
+func (h *trienodeHistory) typ() historyType {
+	return typeTrienodeHistory
+}
+
+// forEach implements the history interface, returning an iterator to traverse the
+// state entries in the history.
+func (h *trienodeHistory) forEach() iter.Seq[stateIdent] {
+	return func(yield func(stateIdent) bool) {
+		for _, owner := range h.owners {
+			for _, path := range h.nodeList[owner] {
+				if !yield(newTrienodeIdent(owner, path)) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// encode serializes the contained trie nodes into bytes.
+func (h *trienodeHistory) encode() ([]byte, []byte, []byte, error) {
+	var (
+		buf           = make([]byte, 64)
+		headerSection bytes.Buffer
+		keySection    bytes.Buffer
+		valueSection  bytes.Buffer
+	)
+	binary.Write(&headerSection, binary.BigEndian, h.meta.version) // 1 byte
+	headerSection.Write(h.meta.parent.Bytes())                     // 32 bytes
+	headerSection.Write(h.meta.root.Bytes())                       // 32 bytes
+	binary.Write(&headerSection, binary.BigEndian, h.meta.block)   // 8 byte
+
+	for _, owner := range h.owners {
+		// Fill the header section with offsets at key and value section
+		headerSection.Write(owner.Bytes())                                       // 32 bytes
+		binary.Write(&headerSection, binary.BigEndian, uint32(keySection.Len())) // 4 bytes
+
+		// The offset to the value section is theoretically unnecessary, since the
+		// individual value offset is already tracked in the key section. However,
+		// we still keep it here for two reasons:
+		// - It's cheap to store (only 4 bytes for each trie).
+		// - It can be useful for decoding the trie data when key is not required (e.g., in hash mode).
+		binary.Write(&headerSection, binary.BigEndian, uint32(valueSection.Len())) // 4 bytes
+
+		// Fill the key section with node index
+		var (
+			prevKey   []byte
+			restarts  []uint32
+			prefixLen int
+
+			internalKeyOffset uint32 // key offset for the trie internally
+			internalValOffset uint32 // value offset for the trie internally
+		)
+		for i, path := range h.nodeList[owner] {
+			key := []byte(path)
+			if i%trienodeDataBlockRestartLen == 0 {
```

Comment:
Same with this, I don't understand why we need the block restarts and why the sharedPrefix will be 0 if we chunk